src/perf/lib/ThroughputServer.cpp

```diff
@@ -0,0 +1,193 @@
+        Session.SetAutoCleanup();
+        Session.SetPeerUnidiStreamCount(THROUGHPUT_SERVER_PEER_UNI);
+        Session.SetDisconnectTimeout(THROUGHPUT_DEFAULT_DISCONNECT_TIMEOUT);
+        Session.SetIdleTimeout(THROUGHPUT_DEFAULT_IDLE_TIMEOUT);
+    }
+}
+
+QUIC_STATUS
+ThroughputServer::Init(
+    _In_ int argc,
+    _In_reads_(argc) _Null_terminated_ char* argv[]
+    ) {
+    uint16_t port = THROUGHPUT_DEFAULT_PORT;
+    TryGetValue(argc, argv, "port", &port);
+
+    const char* localAddress = nullptr;
+    if (!TryGetValue(argc, argv, "listen", &localAddress)) {
+        WriteOutput("Server mode must have -listen\n");
+        return QUIC_STATUS_INVALID_PARAMETER;
+    }
+    if (!ConvertArgToAddress(localAddress, port, &Address)) {
+        WriteOutput("Failed to decode IP address: '%s'!\nMust be *, a IPv4 or a IPv6 address.\n", localAddress);
+        return QUIC_STATUS_INVALID_PARAMETER;
+    }
+
+    TryGetValue(argc, argv, "connections", &NumberOfConnections);
+
+    QUIC_STATUS Status = SecurityConfig.Initialize(argc, argv, Registration);
+
+    if (QUIC_FAILED(Status)) {
+        return Status;
+    }
+
+    return QUIC_STATUS_SUCCESS;
+}
+
+QUIC_STATUS
+ThroughputServer::Start(
+    _In_ QUIC_EVENT StopEvent
+    ) {
+    QUIC_STATUS Status =
+        Listener.Start(
+            &Address,
+            [](HQUIC Handle, void* Context, QUIC_LISTENER_EVENT* Event) -> QUIC_STATUS {
+                return ((ThroughputServer*)Context)->ListenerCallback(Handle, Event);
+            },
+            this);
+    if (QUIC_FAILED(Status)) {
+        return Status;
+    }
+    RefCount = CountHelper{StopEvent};
+    if (NumberOfConnections > 0) {
+        for (uint32_t i = 0; i < NumberOfConnections; i++) {
+            RefCount.AddItem();
+        }
+    } else {
+        //
+        // Add a single item so we can wait on the Count Helper
+        //
+        RefCount.AddItem();
+    }
+    return QUIC_STATUS_SUCCESS;
+}
+
+QUIC_STATUS
+ThroughputServer::Wait(
+    _In_ int Timeout
+    ) {
+    if (Timeout > 0) {
+        RefCount.Wait(Timeout);
+    } else {
+        RefCount.WaitForever();
+    }
+    return QUIC_STATUS_SUCCESS;
+}
+
+QUIC_STATUS
+ThroughputServer::ListenerCallback(
+    _In_ HQUIC /*ListenerHandle*/,
+    _Inout_ QUIC_LISTENER_EVENT* Event
+    ) {
+    QUIC_CONNECTION_CALLBACK_HANDLER Handler;
+    switch (Event->Type) {
+    case QUIC_LISTENER_EVENT_NEW_CONNECTION:
+        Event->NEW_CONNECTION.SecurityConfig = SecurityConfig;
+        Handler =
+            [](HQUIC Conn, void* Context, QUIC_CONNECTION_EVENT* Event) -> QUIC_STATUS {
+                return ((ConnectionData*)Context)->Server->
+                    ConnectionCallback(
+                        Conn,
+                        Event,
+                        (ConnectionData*)Context);
+            };
+        MsQuic->SetCallbackHandler(
+            Event->NEW_CONNECTION.Connection,
+            (void*)Handler,
+            new ConnectionData{ this });
```

Comment:
Can this not be combined into a single statement, like you've done other places?

Reply:
Already done.